sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/splittabledofn/ByteKeyRangeTracker.java

```diff
@@ -0,0 +1,159 @@
+   *
+   * @return {@code true} if the key was successfully claimed, {@code false} if it is outside the
+   *     current {@link ByteKeyRange} of this tracker (in that case this operation is a no-op).
+   */
+  @Override
+  protected synchronized boolean tryClaimImpl(ByteKey key) {
+    checkArgument(
+        lastAttemptedKey == null || key.compareTo(lastAttemptedKey) > 0,
+        "Trying to claim key %s while last attempted was %s",
+        key,
+        lastAttemptedKey);
+    checkArgument(
+        key.compareTo(range.getStartKey()) > -1,
+        "Trying to claim key %s before start of the range %s",
+        key,
+        range);
+    lastAttemptedKey = key;
+    // No respective checkArgument for i < range.to() - it's ok to try claiming keys beyond
+    if (!range.getEndKey().isEmpty() && key.compareTo(range.getEndKey()) > -1) {
+      return false;
+    }
+    lastClaimedKey = key;
+    return true;
+  }
+
+  /**
+   * Marks that there are no more keys to be claimed in the range.
+   *
+   * <p>E.g., a {@link DoFn} reading a file and claiming the key of each record in the file might
+   * call this if it hits EOF - even though the last attempted claim was before the end of the
+   * range, there are no more keys to claim.
+   */
+  public synchronized void markDone() {
+    lastAttemptedKey = range.getEndKey();
+  }
+
+  @Override
+  public synchronized void checkDone() throws IllegalStateException {
+    checkState(
+        next(lastAttemptedKey).compareTo(range.getEndKey()) > -1,
+        "Last attempted key was %s in range %s, claiming work in [%s, %s) was not attempted",
+        lastAttemptedKey,
+        range,
+        next(lastAttemptedKey),
+        range.getEndKey());
+  }
+
+  @Override
+  public String toString() {
+    return MoreObjects.toStringHelper(this)
+        .add("range", range)
+        .add("lastClaimedKey", lastClaimedKey)
+        .add("lastAttemptedKey", lastAttemptedKey)
+        .toString();
+  }
+
+  // Utility methods
+  /**
+   * Calculates the next {@link ByteKey} for a given key by incrementing by one using byte
+   * arithmetic. If the input key is empty it assumes it is a lower bound and returns the 00 byte
+   * array.
+   */
+  @VisibleForTesting
+  static ByteKey next(ByteKey key) {
+    return key.isEmpty()
+        ? ByteKey.of(0)
+        : ByteKey.copyFrom(unsignedCopyAndIncrement(key.getBytes()));
+  }
+
+  // Following methods cloned from org.apache.hadoop.hbase.util.Bytes
+  /**
+   * @return The value of the input incremented by one using byte arithmetic. It treats the input
+   *     byte[] as an unsigned series of bytes, most significant bits first.
+   */
+  private static byte[] unsignedCopyAndIncrement(byte[] input) {
+    byte[] copy = Arrays.copyOf(input, input.length);
+    if (copy == null) {
+      throw new IllegalArgumentException("cannot increment null array");
+    } else {
+      for (int i = copy.length - 1; i >= 0; --i) {
+        if (copy[i] != -1) {
```

Comment:
The number "-1" doesn't look very good in a function called "unsigned..." - are you sure this is correct? Is there a way to make it more obviously correct?

Reply:
Well -1 == 0xff so I am pretty sure there should not be an issue, I will change it if you think is more readable.

Reply:
My worry here was that due to a widening conversion from byte to int we might end up with a comparison of 0xff against 0xffffffff which would fail.